# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/angle/checkout/src/libANGLE/renderer/d3d/d3d11/DebugAnnotator11.cpp
# Commit: 13282d7a47a5
# Full Hash: 13282d7a47a57a73be8d060efd9645b4d642f03e
# Author: Miko Mynttinen <mikokm@gmail.com>
# Date: 2019-09-08 09:46:52
# Regressor Bug: 1578576
# File Overlap Count: 4
# Description:
#   Bug 1578576 - Part 2: Update to ANGLE 3865 r=jgilbert
#   
#   Depends on D44561
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D44579
# ==============================================================================

diff -r 3064469c073d -r 13282d7a47a5 gfx/angle/checkout/src/libANGLE/renderer/d3d/d3d11/DebugAnnotator11.cpp
--- a/gfx/angle/checkout/src/libANGLE/renderer/d3d/d3d11/DebugAnnotator11.cpp	Sat Sep 07 13:36:46 2019 +0000
+++ b/gfx/angle/checkout/src/libANGLE/renderer/d3d/d3d11/DebugAnnotator11.cpp	Sat Sep 07 13:38:36 2019 +0000
@@ -8,119 +8,79 @@
 
 #include "libANGLE/renderer/d3d/d3d11/DebugAnnotator11.h"
 
-#include "common/debug.h"
 #include "libANGLE/renderer/d3d/d3d11/renderer11_utils.h"
 
+#include <versionhelpers.h>
+
 namespace rx
 {
 
-DebugAnnotator11::DebugAnnotator11()
-    : mInitialized(false), mD3d11Module(nullptr), mUserDefinedAnnotation(nullptr)
-{
-    // D3D11 devices can't be created during DllMain.
-    // We defer device creation until the object is actually used.
-}
+DebugAnnotator11::DebugAnnotator11() {}
 
-DebugAnnotator11::~DebugAnnotator11()
-{
-    if (mInitialized)
-    {
-        SafeRelease(mUserDefinedAnnotation);
-
-#if !defined(ANGLE_ENABLE_WINDOWS_STORE)
-        FreeLibrary(mD3d11Module);
-#endif  // !ANGLE_ENABLE_WINDOWS_STORE
-    }
-}
+DebugAnnotator11::~DebugAnnotator11() {}
 
 void DebugAnnotator11::beginEvent(const char *eventName, const char *eventMessage)
 {
-    initializeDevice();
-
     angle::LoggingAnnotator::beginEvent(eventName, eventMessage);
     if (mUserDefinedAnnotation != nullptr)
     {
         std::mbstate_t state = std::mbstate_t();
         std::mbsrtowcs(mWCharMessage, &eventMessage, kMaxMessageLength, &state);
+        std::lock_guard<std::mutex> lock(mAnnotationMutex);
         mUserDefinedAnnotation->BeginEvent(mWCharMessage);
     }
 }
 
 void DebugAnnotator11::endEvent(const char *eventName)
 {
-    initializeDevice();
-
     angle::LoggingAnnotator::endEvent(eventName);
     if (mUserDefinedAnnotation != nullptr)
     {
+        std::lock_guard<std::mutex> lock(mAnnotationMutex);
         mUserDefinedAnnotation->EndEvent();
     }
 }
 
 void DebugAnnotator11::setMarker(const char *markerName)
 {
-    initializeDevice();
-
     angle::LoggingAnnotator::setMarker(markerName);
     if (mUserDefinedAnnotation != nullptr)
     {
         std::mbstate_t state = std::mbstate_t();
         std::mbsrtowcs(mWCharMessage, &markerName, kMaxMessageLength, &state);
+        std::lock_guard<std::mutex> lock(mAnnotationMutex);
         mUserDefinedAnnotation->SetMarker(mWCharMessage);
     }
 }
 
 bool DebugAnnotator11::getStatus()
 {
-#if defined(ANGLE_ENABLE_WINDOWS_STORE)
-    static_assert(NTDDI_VERSION >= NTDDI_WIN10, "GetStatus only works on Win10 and above");
-    initializeDevice();
-
     if (mUserDefinedAnnotation != nullptr)
     {
+        std::lock_guard<std::mutex> lock(mAnnotationMutex);
         return !!(mUserDefinedAnnotation->GetStatus());
     }
 
-    return true;  // Default if initializeDevice() failed
-#else
-    // We can't detect GetStatus() on desktop ANGLE builds so always return true.
-    return true;
-#endif  // ANGLE_ENABLE_WINDOWS_STORE
+    return false;
 }
 
-void DebugAnnotator11::initializeDevice()
+void DebugAnnotator11::initialize(ID3D11DeviceContext *context)
 {
-    if (!mInitialized)
+    // ID3DUserDefinedAnnotation.GetStatus only works on Windows10 or greater.
+    // Returning true unconditionally from DebugAnnotator11::getStatus() means
+    // writing out all compiled shaders to temporary files even if debugging
+    // tools are not attached. See rx::ShaderD3D::prepareSourceAndReturnOptions.
+    // If you want debug annotations, you must use Windows 10.
+    if (IsWindows10OrGreater())
     {
-#if !defined(ANGLE_ENABLE_WINDOWS_STORE)
-        mD3d11Module = LoadLibrary(TEXT("d3d11.dll"));
-        ASSERT(mD3d11Module);
-
-        PFN_D3D11_CREATE_DEVICE D3D11CreateDevice =
-            (PFN_D3D11_CREATE_DEVICE)GetProcAddress(mD3d11Module, "D3D11CreateDevice");
-        ASSERT(D3D11CreateDevice != nullptr);
-#endif  // !ANGLE_ENABLE_WINDOWS_STORE
-
-        ID3D11Device *device         = nullptr;
-        ID3D11DeviceContext *context = nullptr;
-
-        HRESULT hr = E_FAIL;
-
-        // Create a D3D_DRIVER_TYPE_NULL device, which is much cheaper than other types of device.
-        hr = D3D11CreateDevice(nullptr, D3D_DRIVER_TYPE_NULL, nullptr, 0, nullptr, 0,
-                               D3D11_SDK_VERSION, &device, nullptr, &context);
-        ASSERT(SUCCEEDED(hr));
-        if (SUCCEEDED(hr))
-        {
-            mUserDefinedAnnotation =
-                d3d11::DynamicCastComObject<ID3DUserDefinedAnnotation>(context);
-            ASSERT(mUserDefinedAnnotation != nullptr);
-            mInitialized = true;
-        }
-
-        SafeRelease(device);
-        SafeRelease(context);
+        mUserDefinedAnnotation.Attach(
+            d3d11::DynamicCastComObject<ID3DUserDefinedAnnotation>(context));
     }
 }
 
+void DebugAnnotator11::release()
+{
+    mUserDefinedAnnotation.Reset();
+}
+
 }  // namespace rx